### PR TITLE
[5.8] Add `countBy` method to Collection

### PR DESCRIPTION
**Pull request by jamierumbelow**
Adds a `countBy` method to `Illuminate\Support\Collection`.

By default, this method counts each appearance of a given value:

    collect([ 1, 2, 2, 2, 3 ])->countBy(); // == collect([ 1 => 1, 2 => 3, 3 => 1 ])

It also accepts a predicate:

    collect([ 1, 2, 2, 3, 4 ])->countBy(function ($i) {
        return $i % 2 === 0;
    }); // == collect([ true => 3, false => 2 ])

    collect([ 'alice@gmail.com', 'bob@yahoo.fr', 'carlos@gmail.com' ])
        ->countBy(function ($email) {
            return substr(strrchr($email, "@"), 1);
        }); // == collect([ 'gmail.com' => 2, 'yahoo.fr' => 1 ])

### CODE REVIEW

**Comment by driesvints** [comment]
Please fix the StyleCI warnings.

**Review by driesvints — DISMISSED** [review]
Please fix the StyleCI warnings.

File: src/Illuminate/Support/Collection.php [hunk 1]
```diff
@@ -1890,6 +1890,28 @@ public function count()
...
+    /**
+     * Count the number of items in the collection by some predicate.
+     *
+     * @param callable|null
+     * @return static
+     */
+    public function countBy($predicate = null)
+    {
+        if (is_null($predicate)) {
+            $predicate = function ($val, $key) {
```
Comment (BrandonSurowiec) [inline src/Illuminate/Support/Collection.php]:
```suggestion
            $predicate = function ($val) {
```

You aren't utilizing the `$key` so it doesn't need to be set.

**Comment by vlakoff** [comment]
For the "without predicate" mode, couldn't we use [`array_count_values()`](http://php.net/manual/en/function.array-count-values.php)? Probably much, muuch faster.

**Pull request merged**